<h1>PIALANE, ELVIRA J.</h1>
<h3>BSIT-3A</h3>
<h3>ACTIVITY 2</h3>
<HR></HR>

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import f_oneway 

In [2]:
roads_accident = pd.read_csv("datasets\\uk_road_accident.csv")

<hr>

<H1>1. What is the distribution of accident severity?</H1>

In [3]:
roads_accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>Insight: most accidents are classed as Slight (563,801 cases), while Serious (88,217) and Fatal (8,661) accidents are much less common. This tells me that although road accidents happen a lot, they usually aren’t very severe.</h3>

<h1>2. What are the top 5 most common road surface conditions during accidents?</h1>

In [4]:
roads_accident['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     447821
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

<h3>Insight: Most accidents happen on dry roads (447,821), followed by wet roads (186,788). Fewer accidents occur on icy (18,517), snowy (5,890), or flooded roads (1,817). This means accidents happen more often in normal road conditions, likely because more people are driving then.</h3>

<h1>3.What are the most frequent weather conditions during accidents?</h1>

In [5]:
roads_accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>Insight: Most accidents happen in good weather (520,885). Even when it’s raining (79,696), fewer accidents occur. Very few accidents happen during snow, fog, or strong winds. This shows that bad weather isn’t the main cause, people still crash even when the weather is fine.</h3>

<h1>4.What is the average number of casualties per accident?</h1>

In [6]:
roads_accident['Number_of_Casualties'].mean()

1.357040257068864

<h3>Insight: On average, there are about 1.36 casualties per accident. This means that most accidents hurt only one or two people, not large groups.</h3>

<h1>5.What is the trend of accidents over the years?</h1>

In [7]:
roads_accident['Accident_Date'] = pd.to_datetime(roads_accident['Accident Date'], errors='coerce')
roads_accident['Year'] = roads_accident['Accident_Date'].dt.year
roads_accident.groupby('Year').size()

Year
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
dtype: int64

<h3>Insight: The number of accidents has been going down every year: from 71,867 in 2019 to 56,085 in 2022. This drop might be due to less travel, better roads, or safer driving</h3>

<h1>6.Which day of the week has the most accidents?</h1>

In [8]:
roads_accident['Day_of_Week'] = roads_accident['Accident_Date'].dt.day_name()
roads_accident['Day_of_Week'].value_counts().sort_index()

Day_of_Week
Friday       38511
Monday       35715
Saturday     37751
Sunday       37772
Thursday     38009
Tuesday      39701
Wednesday    37548
Name: count, dtype: int64

<h1>7.How many accidents occurred during daylight versus darkness?</h1>

In [9]:
roads_accident['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>Insight: Most accidents happened during the day (484,880). Fewer accidents happened at night with streetlights (129,335). This means more people crash during the day, likely because more people are on the road.</h3>

<h1>8.What is the Pearson R Correlation between the number of casualties and the number of vehicles?</h1>

In [10]:
correlation, p_value = f_oneway (roads_accident['Number_of_Casualties'], roads_accident['Number_of_Vehicles'])
print(f"PRC: {correlation:.4f}")
print(f"P-value: {p_value:.4f}")

PRC: 124644.0567
P-value: 0.0000


<h3>Insight: There is a strong link between the number of vehicles and the number of people hurt in accidents. More vehicles = more casualties. The p-value is 0, which means this result is statistically important.</h3>

<h1>9.How do casualties differ in urban vs. rural areas?</h1>

In [11]:
roads_accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()

Urban_or_Rural_Area
Rural          353515
Unallocated        13
Urban          543024
Name: Number_of_Casualties, dtype: int64

<h3>Insight: Most casualties happened in urban areas (543,824), while rural areas had much fewer (35,315). Cities have more accidents likely because they have more traffic and people.</h3>

<h1>10.Which District Area has the highest number of accidents?</h1>

In [12]:
roads_accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h3>Insight: Birmingham has the highest number of road accidents, with 13,491 recorded incidents. This suggests it may have heavier traffic, denser population, or more high-risk roads compared to other areas like Leeds (8,898) and Manchester (6,722).</h3>

<h1>11.What is the average number of vehicles involved in accidents on different road types?</h1>

In [13]:
roads_accident.groupby('Road_Type')['Number_of_Vehicles'].mean().sort_values(ascending=False)

Road_Type
Dual carriageway      2.009766
Slip road             1.917483
Roundabout            1.901323
Single carriageway    1.795911
One way street        1.585441
Name: Number_of_Vehicles, dtype: float64

<h3>Insight: On average, accidents on dual carriageways involve the most vehicles, with 2.01 vehicles per accident. This is followed by slip roads (1.92) and roundabouts (1.90). Accidents on one-way streets involve the fewest vehicles on average (1.58), possibly due to more controlled traffic flow.</h3>

<h1>12.What is the most common Vehicle Type involved in Fatal accidents?</h1>

In [14]:
fatal_accidents = roads_accident[roads_accident['Accident_Severity'] == 'Fatal']
fatal_accidents['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      6577
Van / Goods 3.5 tonnes mgw or under       467
Motorcycle over 500cc                     339
Bus or coach (17 or more pass seats)      325
Goods 7.5 tonnes mgw and over             216
Motorcycle 125cc and under                189
Taxi/Private hire car                     155
Motorcycle over 125cc and up to 500cc     105
Motorcycle 50cc and under                  95
Other vehicle                              70
Goods over 3.5t. and under 7.5t            67
Minibus (8 - 16 passenger seats)           29
Agricultural vehicle                       21
Pedal cycle                                 6
Name: count, dtype: int64

<h3>Insight: Cars are the most commonly involved vehicle type in fatal accidents, with 6,577 cases. This is far higher than any other type, with the next being vans (467) and motorcycles over 500cc (436). This reflects the high number of cars on the road, increasing their exposure to fatal incidents.</h3>

<h1>13.Which 'District Area' has the highest average 'Number of Casualties'?</h1>

In [15]:
roads_accident.groupby('District Area')['Number_of_Casualties'].mean().sort_values(ascending=False)

District Area
Blaeu Gwent               1.693694
West Somerset             1.613990
Derwentside               1.606186
Chester-le-Street         1.603306
Durham                    1.580431
                            ...   
Kensington and Chelsea    1.136120
Westminster               1.134069
Camden                    1.132783
Islington                 1.126082
City of London            1.126044
Name: Number_of_Casualties, Length: 422, dtype: float64

<h3>Insight: The district Blaenau Gwent has the highest average number of casualties per accident (1.69 casualties on average). Other districts like West Somerset and Derwentside also have higher averages, indicating that when accidents happen in these areas, they tend to be more severe in terms of injuries.</h3>

<h1>14.What is the most common Road Surface Conditions for Serious accidents?</h1>

In [16]:
serious_accidents = roads_accident[roads_accident['Accident_Severity'] == 'Serious']
serious_accidents['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     61638
Wet or damp             23785
Frost or ice             2007
Snow                      565
Flood over 3cm. deep      152
Name: count, dtype: int64

<h3>Insight: The majority of serious accidents occur on dry roads, with 61,638 incidents. The next most common condition is wet or damp roads (23,785). This suggests that serious accidents aren't only caused by poor road conditions even good road conditions (dry) can still lead to serious outcomes, likely due to speed, driver behavior, or other factors.</h3>

<h1>15.How does Light Conditions affect Accident Severity?</h1>

In [17]:
roads_accident.groupby('Light_Conditions')['Accident_Severity'].value_counts().unstack().fillna(0)

Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown     68      794    5622
Darkness - lights lit         1860    19130  108345
Darkness - lights unlit         45      360    2138
Darkness - no lighting        1612     7174   28651
Daylight                      5076    60759  419045

<h3>Insight: Most accidents happen during the day, including many fatal ones:


5,076 fatal

60,759 serious

419,045 slight accidents.

But when it’s dark and there’s no street lighting, the chance of fatal accidents is higher compared to other light conditions. This means poor lighting makes accidents more dangerous.</h3>

<h1>16.What is the most common Vehicle Type involved in accidents in Urban areas?</h1>

In [18]:
urban_accidents = roads_accident[roads_accident['Urban_or_Rural_Area'] == 'Urban']
urban_accidents['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      316053
Van / Goods 3.5 tonnes mgw or under       22003
Bus or coach (17 or more pass seats)      16850
Motorcycle over 500cc                     16700
Goods 7.5 tonnes mgw and over             11151
Motorcycle 125cc and under                10246
Taxi/Private hire car                      8612
Motorcycle over 125cc and up to 500cc      4982
Motorcycle 50cc and under                  4893
Goods over 3.5t. and under 7.5t            3864
Other vehicle                              3643
Agricultural vehicle                       1272
Minibus (8 - 16 passenger seats)           1258
Pedal cycle                                 127
Data missing or out of range                  6
Ridden horse                                  3
Name: count, dtype: int64

<h3>Insight: The most common vehicle involved in accidents in urban areas is cars, with 316,053 accidents. Vans, buses/coaches, and motorcycles also appear frequently but in much smaller numbers compared to cars.</h3>

<h1>17.Are there any specific Vehicle Type and Weather Conditions combinations that are more prone to accidents?</h1>

In [19]:
roads_accident.groupby(['Vehicle_Type', 'Weather_Conditions']).size().sort_values(ascending=False)

Vehicle_Type                          Weather_Conditions   
Car                                   Fine no high winds       392668
                                      Raining no high winds     59940
Van / Goods 3.5 tonnes mgw or under   Fine no high winds        26877
Bus or coach (17 or more pass seats)  Fine no high winds        20398
Motorcycle over 500cc                 Fine no high winds        20245
                                                                ...  
Minibus (8 - 16 passenger seats)      Snowing + high winds          2
Pedal cycle                           Other                         2
Data missing or out of range          Fine + high winds             1
Pedal cycle                           Fine + high winds             1
Ridden horse                          Raining no high winds         1
Length: 115, dtype: int64

<h3>Insight: The most common combination for accidents is cars in fine weather with no high winds, with 392,668 accidents. This suggests that most accidents happen in good weather conditions, possibly because more vehicles are on the road.</h3>

<h1>18.What is the distribution of Accident Severity for accidents on Single carriageway?</h1>

In [20]:
single_carriageway_accidents = roads_accident[roads_accident['Road_Type'] == 'Single carriageway']
single_carriageway_accidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     415557
Serious     70059
Fatal        6527
Name: count, dtype: int64

<h3>Insight: For single carriageway roads:

Slight accidents are the most common with 415,557 cases.

Serious accidents total 70,059.

Fatal accidents are fewer, with 6,527 cases.</h3>

<H1>For Insight No. 19 and 20</H1>

In [21]:
severity_weather_road = roads_accident.groupby(['Accident_Severity', 'Road_Surface_Conditions'],observed=False).size()
severity_weather_road.unstack()

Road_Surface_Conditions     Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Accident_Severity                                                           
Fatal                      5788                    23           193    35   
Serious                   61638                   152          2007   565   
Slight                   380395                   842         16317  5290   

Road_Surface_Conditions  Wet or damp  
Accident_Severity                     
Fatal                           2620  
Serious                        23785  
Slight                        160303

<h3>Insight:

19. Most accidents happen on dry roads, with the highest number of slight accidents (380,395), serious accidents (61,638), and fatal accidents (5,788). This shows that dry conditions still see many accidents, likely due to higher traffic volumes.
20. Wet or damp roads lead to a significant number of serious accidents (23,785) and fatal accidents (2,620), highlighting that slippery surfaces increase the risk of more severe crashes compared to conditions like snow or flooding.</h3>

<h1>Discussion</h1>
<h3>Adjusting Data Types</h3>
<hr>

In [22]:
roads_accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date                      object
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
Accident_Date              datetime64[ns]
Year                              float64
Day_of_Week                        object
dtype: object

<h1>Extracting date information using pandas date time</h1>

In [34]:
roads_accident['Accident Date'] = pd.to_datetime(roads_accident['Accident Date'],dayfirst = True, errors = 'coerce')
roads_accident['Year'] = roads_accident ['Accident Date'].dt.year
roads_accident['Month'] = roads_accident ['Accident Date'].dt.month
roads_accident['Day'] = roads_accident['Accident Date'].dt.day
roads_accident['DayOfWeek'] = roads_accident['Accident Date'].dt.dayofweek
roads_accident['Accident Date'].value_counts()

Accident Date
2021-02-11    685
2021-10-06    680
2019-12-06    678
2019-02-01    676
2021-12-04    667
             ... 
2022-01-09    236
2022-12-02    236
2022-04-05    231
2022-02-07    213
2022-01-10    123
Name: count, Length: 576, dtype: int64

In [25]:
roads_accident.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date              395672
Latitude                       25
Light_Conditions                0
District Area                   0
Longitude                      26
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions       726
Road_Type                    4520
Urban_or_Rural_Area            15
Weather_Conditions          14128
Vehicle_Type                    0
Accident_Date              395672
Year                            0
Day_of_Week                395672
dtype: int64